### PR TITLE
[FIX] website, website_form_project: prevent removal of e-mail field

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] when the e-mail field of the "Create Task" forms have been turned into a required custom field, nothing prevents the user from removing it from the form.
However this field is still required by the controller that receives the form's submit. This leads to errors when such a form is submitted and no e-mail field was present.

This commit introduces a `requiredController` attribute on fields, which marks the field with the same class as `modelRequired` fields so that its deletion is prevented.

[1]: https://github.com/odoo/odoo/commit/dede76bd5c898c7b68b1b70c44bb73588c2f594e

opw-4243770

### CODE REVIEW

**Comment by qsm-odoo** [comment]
@robodoo delegate=xO-Tx 
@xO-Tx

File: addons/website/static/src/snippets/s_website_form/options.js [hunk 1]
```diff
@@ -242,6 +246,9 @@ const FieldEditor = FormEditor.extend({
         const labelText = this.$target.find('.s_website_form_label_content').text();
         if (this._isFieldCustom()) {
             field = this._getCustomField(this.$target[0].dataset.type, labelText);
+            if (this.$target[0].classList.contains("s_website_form_model_required")) {
+                field.controllerRequired = true;
+            }
```
Comment (xO-Tx) [inline addons/website/static/src/snippets/s_website_form/options.js:251]: I don't understand this diff, I think the idea here is to ensure the field state consistency, but there is already a code in `_setActiveProperties()` that will force the field to be `modelRequired` anyway :thinking:
```javascript
field.modelRequired = classList.contains('s_website_form_model_required');
```

**Review by xO-Tx — CHANGES_REQUESTED** [review]
@bso-odoo Some remarks about this fix (for tracking purposes):

1. In `17.0`, removing the field won't cause any issue, since the bug was caused by this feature from 17.4: https://github.com/odoo/odoo/commit/1aa5cbb06be85e94c01f8f55ef57b9415b9bb50f (it mainly uses the field to compare the default values with submitted ones and send a message on the record...). But yes, I agree it's better to have the feature from 17.0 (also to keep the same behavior from before 17.0) :+1: 

2. Also, I don't think `controllerRequired` fields should be cloned: you will end up with multiple "unremovable" fields :grimacing:

File: addons/website/static/src/snippets/s_website_form/options.js (hunk 1 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_website_form/options.js:251]: Without this, after save, when re-selecting the field, the `controllerRequired` information is lost.

**Comment by bso-odoo** [comment]
@xO-Tx Added a selector update to prevent the duplication of the "controller required" fields.

File: addons/website/static/src/snippets/s_website_form/options.js (hunk 1 above)
Comment (xO-Tx) [reply addons/website/static/src/snippets/s_website_form/options.js:251]: Yes, but don't you find it very confusing :thinking:  ? We use the same class
`"s_website_form_model_required"` for `modelRequired` and `controllerRequired` which means every `modelRequired` custom field will be `controllerRequired` and every `controllerRequired` field will be `modelRequired`… So, in the end, isn't equivalent to setting `modelRequired = true` on the field (or a global attribute to say that the field is required in the backend) ?

Comment (bso-odoo) [reply addons/website/static/src/snippets/s_website_form/options.js]: Now using dedicated class [(diff)](https://github.com/odoo/odoo/compare/5b7af3730f04f7c77c60a87dbfa633c43db1069e..abf6ca6b29ba7e86ce9caaaceb624f971a396c72)

**Comment by xO-Tx** [comment]
Closed, since the "missing email field issue" was fixed in `17.4` (where the error occurs) by allowing forms without emails to be submitted. See: https://github.com/odoo/odoo/pull/185032 :+1:
cc @bso-odoo

**Pull request closed without merging**